### ISSUE TITLE
[Feature Request] Better bookmark organization

### ISSUE THREAD

**Issue by matt-gardner** [scrooloose/nerdtree#495]
I find this plugin to be incredibly useful to me as I write software.  However, I have lots of projects going on at a time, and I would like to have bookmarks to various places in the code for each of these projects.  For example, one project has code in both java and scala, and I have 5 bookmarks to various places in the code: the base directory, for easy access to `build.sbt` and related files, the main directory containing scala code (bypassing the first few levels of the package hierarchy), the main directory containing java code, and the test directories for both java and scala.  If you multiply this by several projects, the number of bookmarks quickly becomes unmanageable.  Being able to hide the bookmarks in the tree is nice, but it's not really ideal, because you still get a huge list to sort through.

It seems like there are two possible solutions to this:

(1) Allow for project-specific bookmarks.  That is, NERDTree would look recursively in the filesystem for a directory containing a .NERDTreeBookmarks file, and load those bookmarks (perhaps in addition to the global bookmarks in the plugin directory).  This would reduce clutter from other projects when I'm working on a specific one.  The only thing that seems tricky about this is when creating bookmarks, you need to specify where in the filesystem they should live.

(2) Allow for nesting of bookmarks in a tree structure, similar to how the filesystem is displayed.  For instance, if I specify bookmarks like project1/home, project1/scala-main and project1/scala-test, NERDTree could create a nested structure for project1/, which, when opened, displays the bookmarks below it.  You can currently create bookmarks with these names without any trouble, so the only thing to be added here would be a tree-format option in the display.

I can try to take a stab at implementing one of these, though it'd be nice to have some input on whether others think this idea is useful, and which option is better, before I start trying.

**Comment by petoalbert** [comment]
A solution to this problem would be very useful for me, too. 

I would choose the second way that you described, that is, keeping every bookmark in the same file. This looks more flexible for me because it avoids the need of specifying the right bookmarks file, as you mentioned, and, more generally, because there can be related sets of bookmarks where one could not choose a specific directory which would really describe them or be common to them. Interrelated projects is a good example to this. (In that case, either one of the related directories should be chosen to contain the bookmarks file, or a specific "common" directory could be created for that purpose, but I find both methods inconvenient.)

There is another thing, which I think, would fit in this issue, or could be regarded as a third solution: Sometimes, when working on big projects with deep directory structure, I need to jump between related files in some specific subdirectories. For example, where the project is build up from modules and I have to read the source code of two interacting modules. Adding bookmarks would solve the problem, but there are cases where the directories of interest quickly become obsolete (for example, in one hour). In a short time I would have to edit different parts of the code and those bookmarks would become irrelevant. Instead of having to delete those obsolete bookmarks, I would be happy to be able to create temporary bookmarks, which don't get saved to a file and live only for the lifetime of the NERDTree instance.

**Comment by scrooloose** [comment]
Another option is to have one bookmark per project and filter out everything except the 5 dirs that you care about. I have been playing around with this today (for proof of concept) and have made it work quite well.

Here's an example to demo what I mean:

I have a rails project, and I only care about `./app/views/layouts/**` and `./public/system/**`. I create some filters for the project so that when I open a nerdtree in the project root I see this:

![first-open](https://cloud.githubusercontent.com/assets/1671/11314219/921eba2e-8fdb-11e5-90c5-8fa2aa551881.png)

And if I open some dirs I see this:

![open-dirs](https://cloud.githubusercontent.com/assets/1671/11314225/9ea18ef2-8fdb-11e5-8c3b-a2110e24937e.png)

At this point, if I were to deactivate the file filters (by pressing `f`) then I'd see all files in the project:

![deactivate-filefilters](https://cloud.githubusercontent.com/assets/1671/11314244/bbc03eb6-8fdb-11e5-84fc-b9a22f812af3.png)

The code for this is fairly straight forward: https://gist.github.com/scrooloose/f832454871d6bcfb52a4

If this solution is interesting then I'd polish it up.

I'm in the process of creating a project nerdtree plugin that is intended to replace bookmarks in the long run (for #248). Filtering irrelevant dirs out of the tree sounds like exactly the kind of thing it should do.

Thoughts?

**Comment by matt-gardner** [comment]
That approach seems useful for some cases, but not as much for me particularly.  I do very much like the idea of projects, and some of the other issues you're working on that for.

My problem with this solution for this problem is that it doesn't bypass the directory structure.  For instance, if I have files like this:

> src
>> main
>>> org
>>>> allenai
>>>>> parser
>>>>>> [interesting files here, with more directories]
>> test
>>> org
>>>> allenai
>>>>> parser
>>>>>> [interesting files here, with more directories]

It's not that I want to filter anything out.  I just don't have the horizontal real estate on my screen to have that many directories open in the nerdtree window, so I have bookmarks to the "parser" directory under both main and test, and switch between the bookmarked directories as necessary.  So, an alternative that would be somewhat similar to what you suggested but still helpful to me would be a way to collapse some directories in the tree, including their indentation.  If, for example, the structure looked like this:

> src
>> main/org/allenai/parser
>>> [interesting files here]
>> test/org/allenai/parser
>>> [interesting files here]

that would be really helpful.  This is similar to how github automatically combines directories when there is only one subdirectory, though it'd also be nice if there were a way to do this even when there are multiple subdirectories, with a command or something (and a way to undo it, so I can add subdirectories or whatever).

Does that make sense?

**Comment by scrooloose** [comment]
Hey @matt-gardner, yeah I understand.

I have implemented a proof of concept version of this here: https://github.com/scrooloose/nerdtree/tree/shorten_cascading_dirs

Give it a go and let me know what you think. From my brief testing it seems stable... but I'm sure its buggy as hell :)

**Comment by matt-gardner** [comment]
Good work - this seems to be a substantial improvement over the master branch, and by itself is probably enough that I don't need bookmarks any more.  I tried adding subdirectories that make a directory no longer cascadable, and then removing the directory, and the nerdtree window updated like I expected.  I didn't find any bugs in my brief testing of the functionality.

The only thing I can think of that would make this better would be a key mapping to cascade a directory in the nerdtree window even when there are multiple subdirectories, and to undo a cascade.  This would look something like this:

Before keypress:

> src
>> main
>>> [subtree]

>> test
>>> [subtree]

After pressing the key on the src/main directory, or the src/test directory:

> src/main
>> [subtree]

> src/test
>> [subtree]

When pressing the undo key on either the src/main or src/test directory, the UI would go back to how it was.

This seems like it would be harder to implement, and not as universally useful, so do with the feature request as you wish.  Thanks for the improvements so far, they are incredibly helpful.

**Comment by scrooloose** [comment]
OK, I have merged this onto master. The only bug I found/fixed was with flags on cascading dirs which was messing with the git plugin -> https://github.com/Xuyuanp/nerdtree-git-plugin.

As you said, your second feature request is a bit more drastic to implement and probably wouldn't have the pay off that the new cascading dirs have. For now I wont attempt it. I'm prepared to reconsider if more people suggest the same, but for now it seems like we have fixed maybe 80% of your pain.

Thanks for testing, and let me know if there are any issues/suggestions.
